### ISSUE TITLE
Lambda variable renaming

### ISSUE THREAD

**Issue by pdinklag** [HeartPattern/MC-Remapper#4]
There appears to be one thing left preventing remapped and then decompiled sources from compiling. Sorry if that's not a goal of your project, but this may still be interesting to you.

Using `--fixlocalvar=rename`, variables are named according to whatever index ASM assigns to them when visiting them. However, there's a problem with lambdas. Again, I'll use a part of `net.minecraft.world.entity.monster.Creeper` as an example:

```java
/*     */   protected boolean mobInteract(Player debug1, InteractionHand debug2) {
/* 211 */     ItemStack debug3 = debug1.getItemInHand(debug2);
/* 212 */     if (debug3.getItem() == Items.FLINT_AND_STEEL) {
/* 213 */       this.level.playSound(debug1, getX(), getY(), getZ(), SoundEvents.FLINTANDSTEEL_USE, getSoundSource(), 1.0F, this.random.nextFloat() * 0.4F + 0.8F);
/* 214 */       if (!this.level.isClientSide) {
/* 215 */         ignite();
/* 216 */         debug3.hurtAndBreak(1, (LivingEntity)debug1, debug1 -> debug1.broadcastBreakEvent(debug0));
/*     */       } 
/* 218 */       return true;
/*     */     } 
/* 221 */     return super.mobInteract(debug1, debug2);
/*     */   }
```
Line 216 can't be compiled for two reasons:

1. The lambda parameter is named `debug1`, but that name already exists in the scope of `mobInteract`.
2. The variable `debug0` doesn't even exist and should be `debug2`, the `InteractionHand` parameter of `mobInteract`.

I tried to have a look at this myself, but I have zero experience with ASM. Using only the index of the local variable to name it is not enough, though. I tried keeping a static counter in a companion object and give each local variable a unique name, but that won't work either.

May this be a limitation or an issue with ASM? The index 0 shouldn't even exist - instead, it should use the local variable `debug2` from the calling scope . Also, there should be a new index for the lambda parameter.

**Comment by HeartPattern** [comment]
I haven't consider lambda and anonymous class. Thanks for reporting. I will try to fix it.

**Comment by pdinklag** [comment]
I did some research. This is purely a decompiler issue, MC-Remapper is fine.

Compiled lambdas are just methods that get called within some `invokevirtual` magic. Their variable names should really never matter at all, because they are not visible in uncompiled code. Instead, they are linked either to the lambda parameters or to variables that are part of a closure.

I've been using [jd-gui](https://github.com/java-decompiler/jd-gui), which displays the lambda's own parameter names. I'm very certain that's a bug. I've now tried [bytecode-viewer](https://github.com/Konloch/bytecode-viewer) instead which uses a different decompilation engine, and that one handles this case just fine and produces recompilable code. It also handles the snowmen fine, so there's not even any need for renaming anymore.

So, sorry for bugging and thanks again for this project!
